src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,460 @@
+from __future__ import annotations
+
+import abc
+import os
+import sys
+from dataclasses import dataclass
+
+from globus_sdk import (
+    AuthClient,
+    AuthLoginClient,
+    ConfidentialAppAuthClient,
+    NativeAppAuthClient,
+    Scope,
+)
+from globus_sdk._types import UUIDLike
+from globus_sdk.authorizers import GlobusAuthorizer
+from globus_sdk.exc import GlobusSDKUsageError
+from globus_sdk.experimental.auth_requirements_error import (
+    GlobusAuthorizationParameters,
+)
+from globus_sdk.experimental.login_flow_manager import (
+    CommandLineLoginFlowManager,
+    LoginFlowManager,
+)
+from globus_sdk.experimental.tokenstorage import JSONTokenStorage, TokenStorage
+from globus_sdk.scopes import AuthScopes
+
+from ._validating_token_storage import ValidatingTokenStorage
+from .authorizer_factory import (
+    AccessTokenAuthorizerFactory,
+    AuthorizerFactory,
+    ClientCredentialsAuthorizerFactory,
+    RefreshTokenAuthorizerFactory,
+)
+
+
+def _default_filename(app_name: str) -> str:
+    r"""
+    construct the filename for the default JSONTokenStorage to use
+
+    on Windows, this is typically
+        ~\AppData\Local\globus\app\{app_name}/tokens.json
+
+    on Linux and macOS, we use
+        ~/.globus/app/{app_name}/tokens.json
+    """
+    if sys.platform == "win32":
+        # try to get the app data dir, preferring the local appdata
+        datadir = os.getenv("LOCALAPPDATA", os.getenv("APPDATA"))
+        if not datadir:
+            home = os.path.expanduser("~")
+            datadir = os.path.join(home, "AppData", "Local")
+        return os.path.join(datadir, "globus", "app", app_name, "tokens.json")
+    else:
+        return os.path.expanduser(f"~/.globus/app/{app_name}/tokens.json")
+
+
+@dataclass(frozen=True)
+class GlobusAppConfig:
+    """
+    Various configuration options for controlling the behavior of a ``GlobusApp``.
+
+    :param login_flow_manager: an optional ``LoginFlowManager`` instance or class.
+        An instance will be used directly when driving app login flows. A class will
+        be initialized with the app's ``login_client`` and this config's
+        ``request_refresh_tokens``.
+        If not given the default behavior will depend on the type of ``GlobusApp``.
+    :param token_storage: a ``TokenStorage`` instance or class that will
+        be used for storing token data. If not passed a ``JSONTokenStorage``
+        will be used.
+    :param request_refresh_tokens: If True, the ``GlobusApp`` will request refresh
+        tokens for long lived access.
+    """
+
+    login_flow_manager: LoginFlowManager | type[LoginFlowManager] | None = None
```

Comment:
This is an elegant solution to my initial thread.

I'm very slightly hesitant about having the constructor interface be mandated like this but I can live with it. If a user is subclassing their own LoginFlowManager, they're 90% going to be pre-instantiating it notgiving it to us to instantiate it.